src/policies/EmissionManager.sol

```diff
@@ -0,0 +1,469 @@
+import {IgOHM} from "interfaces/IgOHM.sol";
+
+import {RolesConsumer, ROLESv1} from "modules/ROLES/OlympusRoles.sol";
+import {TRSRYv1} from "modules/TRSRY/TRSRY.v1.sol";
+import {PRICEv1} from "modules/PRICE/PRICE.v1.sol";
+import {MINTRv1} from "modules/MINTR/MINTR.v1.sol";
+import {CHREGv1} from "modules/CHREG/CHREG.v1.sol";
+
+interface BurnableERC20 {
+    function burn(uint256 amount) external;
+}
+
+interface Clearinghouse {
+    function principalReceivables() external view returns (uint256);
+}
+
+contract EmissionManager is Policy, RolesConsumer {
+    using FullMath for uint256;
+
+    // ========== ERRORS ========== //
+
+    error OnlyTeller();
+    error InvalidMarket();
+    error InvalidCallback();
+    error InvalidParam(string parameter);
+    error CannotRestartYet(uint48 availableAt);
+    error RestartTimeframePassed();
+    error NotActive();
+    error AlreadyActive();
+
+    // ========== EVENTS ========== //
+
+    event SaleCreated(uint256 marketID, uint256 saleAmount);
+    event BackingUpdated(uint256 newBacking, uint256 supplyAdded, uint256 reservesAdded);
+
+    // ========== DATA STRUCTURES ========== //
+
+    struct BaseRateChange {
+        uint256 changeBy;
+        uint48 beatsLeft;
+        bool addition;
+    }
+
+    // ========== STATE VARIABLES ========== //
+
+    /// @notice active base emissions rate change information
+    /// @dev active until beatsLeft is 0
+    BaseRateChange public rateChange;
+
+    // Modules
+    TRSRYv1 public TRSRY;
+    PRICEv1 public PRICE;
+    MINTRv1 public MINTR;
+    CHREGv1 public CHREG;
+
+    // Tokens
+    // solhint-disable const-name-snakecase
+    ERC20 public immutable ohm;
+    IgOHM public immutable gohm;
+    ERC20 public immutable reserve;
+    ERC4626 public immutable sReserve;
+
+    // External contracts
+    IBondSDA public auctioneer;
+    address public teller;
+
+    // Manager variables
+    uint256 public baseEmissionRate;
+    uint256 public minimumPremium;
+    uint48 public vestingPeriod; // initialized at 0
+    uint256 public backing;
+    uint8 public beatCounter;
+    bool public locallyActive;
+    uint256 public activeMarketId;
+
+    uint8 internal _oracleDecimals;
+    uint8 internal immutable _ohmDecimals;
+    uint8 internal immutable _reserveDecimals;
+
+    /// @notice timestamp of last shutdown
+    uint48 public shutdownTimestamp;
+    /// @notice time in seconds that the manager needs to be restarted after a shutdown, otherwise it must be re-initialized
+    uint48 public restartTimeframe;
+
+    uint256 internal constant ONE_HUNDRED_PERCENT = 1e18;
+
+    // ========== SETUP ========== //
+
+    constructor(
+        Kernel kernel_,
+        address ohm_,
+        address gohm_,
+        address reserve_,
+        address sReserve_,
+        address auctioneer_,
+        address teller_
+    ) Policy(kernel_) {
+        // Set immutable variables
+        if (ohm_ == address(0)) revert("OHM address cannot be 0");
```

Comment:
would also remove these checks. Its gas inefficient with no real reason to do so. Its more likely you put a wrong value than you put a 0 here.